### DISCUSSION TITLE
`useState` with Form Controls

### DISCUSSION THREAD

**Discussion by Cryptobyte** [react-hook-form/react-hook-form#7043] [Q&A] [answered]
Hey everyone!

I am really struggling to figure out how to use a state variable that I have defined with `useState` inside a control. Here's some pseudocode that describes the issue, hopefully someone can help!

```
const [num, setNum] = useState(0);

useEffect(() => {
  // API call that updates `num` and runs on load
  // I want to use this updated number in the component by default

}, []);

return (
  // ...

  <FormControl isInvalid={errors.numfield}>
    <FormLabel htmlFor='numfield'>Num field</FormLabel>

    <NumberInput step={1} defaultValue={num}>
      <NumberInputField
        id='numfield'
        type='number' 
        value={num}
        placeholder='Num Field' 
        errorBorderColor='crimson'
        onChange={(e) => setNum(e.target.value)}
        {...register('numfield', { required: 'Num field is required and must be a number', min: 1 })} />

      <NumberInputStepper>
        <NumberIncrementStepper />
        <NumberDecrementStepper />
      </NumberInputStepper>
    </NumberInput>

    <FormErrorMessage>
      {errors.numfield && errors.numfield.message}
    </FormErrorMessage>
  </FormControl>

  // ...
);
```

**Comment by ritikbanger** [comment, marked as answer]
Hello @Cryptobyte, The register method is registering the inputs and will pass the data via handleSubmit handler.
You need to create a button with handleSubmit handler like this:
` <button label='save_state' onClick={handleSubmit(onSubmit)} />`

After that in the onSubmit function you can set the states like this:
`const onSubmit = async (data: any) => {
    setNum({ ...data })
  }`

**Reply by Cryptobyte** [reply]
Yes I understand the onSubmit I am trying to modify data before submission

**Reply by ritikbanger** [reply]
Since the order of execution will be UseEffect and then UseState. The num returned by the API will get updated to the default value '0' that you gave in the UseState, that's why for the first time, the num=0 and not equal to the value returned by the API.

**Reply by Cryptobyte** [reply]
By adding this code I can see that the value of the `useState` variable is being updated however the value never changes in my input component:

`useEffect(() => console.log(num), [num]);`

I'm getting a log for the state being 0 and the state being the updated value from the API despite this the component stays at 0 value and does not change with the state variable. That's kind of what leads me to believe that maybe it's some sort of binding issue? I can't find any good documentation about using this library with useState variables though so I am not sure.

However assuming that the order of execution is the problem how would this he solved? It seems fairly commonplace to load data inside `useEffect` and store in a state variable but maybe there's something wrong with my implementation.

**Reply by ritikbanger** [reply]
I hope you have the useForm in your code, if not try adding this and use num as default value in that only instead of passing in the field.

```
const methods = useForm({
    defaultValues: useMemo(() => {
      return num
    }, []),
    resolver: If you are using any resolver,
  })

  const {
    handleSubmit,
    reset,
    register,
    formState: { errors },
  } = methods

```

**Comment by ritikbanger** [comment]
@bluebill1049 Can you close this?

**Reply by bluebill1049** [reply]
thanks @ritikbanger 👍
